File: src/main/java/com/tngtech/jgiven/util/PsiElementUtil.java
package com.tngtech.jgiven.util;

import com.intellij.psi.PsiElement;

import java.util.Optional;

public class PsiElementUtil {

    @SuppressWarnings("unchecked")
    public static <E extends PsiElement> Optional<E> findParentOfTypeOn(PsiElement element, Class<E> toFind) {
        if (element == null) {
            return Optional.empty();
        }
        if (toFind.isAssignableFrom(element.getClass())) {
            return Optional.of((E) element);
        }
        return findParentOfTypeOn(element.getParent(), toFind);
    }
}

File: src/main/java/com/tngtech/jgiven/util/PsiElementUtil.kt
package com.tngtech.jgiven.util

import com.intellij.psi.PsiElement

inline fun <reified E : PsiElement> findParentOfTypeOn(element: PsiElement?): E? {
    element ?: return null

    var current: PsiElement? = element
    while(current != null) {
        if (current is E) {
            return current
        }
        current = current.parent
    }
    return null
}
